File: PyProject/leetcode/merge.py
from typing import List


def merge(arr: List[int]):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]

        merge(L)
        merge(R)

        i = j = k = 0

        # 合并两个数组
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        while i < len(L):
            arr[k] = L[i]
            k += 1
            i += 1
        while j < len(R):
            arr[k] = R[j]
            k += 1
            j += 1


if __name__ == '__main__':
    a = [1, 4, 3, 2, 8, 3, 6]
    merge(a)
    print(a)

File: GoProject/leetcode/binary_search/merge.go
package main

import "fmt"

func mergeSort(nums []int) []int {
	if len(nums) <= 1 {
		return nums
	}

	result := []int{}

	mid := len(nums) / 2

	left := mergeSort(nums[0:mid])
	right := mergeSort(nums[mid:len(nums)])

	ll := len(left)
	lr := len(right)

	l, r := 0, 0
	for l <= ll && r <= lr {
		if left[l] < right[r] {
			result = append(result, left[l])
			l++
		} else {
			result = append(result, right[r])
			r++
		}
	}
	result = append(result, left[l:]...)
	result = append(result, right[r:]...)

	return result
}

func main() {
	arr := []int{10, 7, 8, 9, 1, 5}
	result := mergeSort(arr)
	fmt.Println(result)
}
